File: go/rrule-gen.go
package main

import (
	"encoding/json"
	"fmt"
	"log"
	"math/rand"
	"net/http"
	"time"

	"github.com/teambition/rrule-go"
)

type interval int

const (
	MINUTE interval = iota
	HOUR
	DAY
	WEEKLY
)

func GetUnixTimeValueForInterval(reqInterval interval) int64 {
	switch reqInterval {
	case MINUTE:
		return 60
	case HOUR:
		return 60 * GetUnixTimeValueForInterval(MINUTE)
	case DAY:
		return 24 * GetUnixTimeValueForInterval(HOUR)
	case WEEKLY:
		return 7 * GetUnixTimeValueForInterval(DAY)
	default:
		return 0
	}
}

func GenerateRandomDatesWithInterval(intervalType interval, amount int64) (time.Time, time.Time) {
	initialTime := rand.Int63n(time.Now().Unix()-94608000) + 94608000
	min := time.Unix(initialTime, 0)
	max := time.Unix(initialTime+GetUnixTimeValueForInterval(intervalType)*amount, 0)
	return min, max
}

func ProjectMaxExpectedEvents() []time.Time {

	sampleRule, _ := rrule.NewRRule(rrule.ROption{
		Freq:      rrule.WEEKLY,
		Dtstart:   time.Date(2021, 2, 1, 10, 30, 0, 0, time.UTC),
		Interval:  1,
		Byweekday: []rrule.Weekday{rrule.MO, rrule.WE, rrule.FR},
	})

	iterations := 1000000

	res := make([]time.Time, 0)

	for i := 0; i < iterations; i++ {
		min, max := GenerateRandomDatesWithInterval(DAY, 1)
		res = append(res, sampleRule.Between(min, max, true)...)
	}

	return res
}

const numberOfEvents = 100000

func ProjectDailies() []time.Time {
	dtstart := time.Date(2021, 2, 1, 10, 30, 0, 0, time.UTC)
	sampleRule, _ := rrule.NewRRule(rrule.ROption{
		Freq:     rrule.DAILY,
		Dtstart:  dtstart,
		Interval: 1,
	})

	return sampleRule.Between(dtstart, dtstart.AddDate(0, 0, numberOfEvents), true)
}

func ProjectMultipleRulesDaily() []time.Time {
	dtstart := time.Date(2021, 2, 1, 10, 30, 0, 0, time.UTC)
	sampleRule, _ := rrule.NewRRule(rrule.ROption{
		Freq:     rrule.DAILY,
		Dtstart:  dtstart,
		Interval: 1,
	})

	res := make([]time.Time, 0)

	for i := 0; i < numberOfEvents; i++ {
		res = append(res, sampleRule.Between(dtstart, dtstart, true)...)
	}

	return res
}

func ProjectMultipleRulesDailyOld() []time.Time {
	dtstart := time.Date(2021, 2, 1, 10, 30, 0, 0, time.UTC)
	sampleRule, _ := rrule.NewRRule(rrule.ROption{
		Freq:     rrule.DAILY,
		Dtstart:  dtstart.AddDate(-2, 0, 0),
		Interval: 1,
	})

	res := make([]time.Time, 0)

	for i := 0; i < numberOfEvents; i++ {
		res = append(res, sampleRule.Between(dtstart, dtstart, true)...)
	}

	return res
}

func ProjectWeekly() []time.Time {
	dtstart := time.Date(2021, 2, 1, 10, 30, 0, 0, time.UTC)
	sampleRule, _ := rrule.NewRRule(rrule.ROption{
		Freq:      rrule.WEEKLY,
		Dtstart:   dtstart,
		Interval:  1,
		Byweekday: []rrule.Weekday{rrule.MO, rrule.WE, rrule.FR},
	})

	return sampleRule.Between(dtstart, dtstart.AddDate(0, 0, numberOfEvents/3*7), true)
}
func ProjectMultipleRulesWeekly() []time.Time {
	dtstart := time.Date(2021, 2, 1, 10, 30, 0, 0, time.UTC)
	sampleRule, _ := rrule.NewRRule(rrule.ROption{
		Freq:      rrule.WEEKLY,
		Dtstart:   dtstart,
		Interval:  1,
		Byweekday: []rrule.Weekday{rrule.MO, rrule.WE, rrule.FR},
	})

	iterations := numberOfEvents / 3

	res := make([]time.Time, 0)

	for i := 0; i < iterations; i++ {
		res = append(res, sampleRule.Between(dtstart, dtstart.AddDate(0, 0, 6), true)...)
	}

	return res
}
func ProjectMultipleRulesWeeklyOld() []time.Time {
	dtstart := time.Date(2021, 2, 1, 10, 30, 0, 0, time.UTC)
	sampleRule, _ := rrule.NewRRule(rrule.ROption{
		Freq:      rrule.WEEKLY,
		Dtstart:   dtstart.AddDate(-2, 0, 0),
		Interval:  1,
		Byweekday: []rrule.Weekday{rrule.MO, rrule.WE, rrule.FR},
	})

	iterations := numberOfEvents / 3

	res := make([]time.Time, 0)

	for i := 0; i < iterations; i++ {
		res = append(res, sampleRule.Between(dtstart, dtstart.AddDate(0, 0, 6), true)...)
	}

	return res
}

func GenerateEvents() []time.Time {
	// 3 days a week
	r1, _ := rrule.NewRRule(rrule.ROption{
		Freq:      rrule.WEEKLY,
		Dtstart:   time.Date(2021, 2, 1, 10, 30, 0, 0, time.UTC),
		Interval:  1,
		Byweekday: []rrule.Weekday{rrule.MO, rrule.WE, rrule.FR},
	})

	// Daily
	r2, _ := rrule.NewRRule(rrule.ROption{
		Freq:    rrule.DAILY,
		Dtstart: time.Date(2021, 2, 1, 10, 30, 0, 0, time.UTC),
	})

	// With exceptions
	set := rrule.Set{}
	r3, _ := rrule.NewRRule(rrule.ROption{
		Freq:    rrule.WEEKLY,
		Dtstart: time.Date(2021, 2, 1, 17, 30, 0, 0, time.UTC),
	})
	set.RRule(r3)
	set.ExDate(time.Date(2021, 8, 16, 17, 30, 0, 0, time.UTC))
	set.ExDate(time.Date(2021, 9, 27, 17, 30, 0, 0, time.UTC))
	set.RDate(time.Date(2021, 8, 5, 10, 30, 0, 0, time.UTC))
	set.RDate(time.Date(2021, 8, 16, 10, 30, 0, 0, time.UTC))
	set.RDate(time.Date(2021, 9, 27, 10, 30, 0, 0, time.UTC))

	res := r1.Between(
		time.Date(2021, 8, 1, 0, 0, 0, 0, time.UTC),
		time.Date(2021, 8, 7, 23, 59, 59, 0, time.UTC), true)
	res = append(res, r2.Between(
		time.Date(2021, 8, 1, 0, 0, 0, 0, time.UTC),
		time.Date(2021, 8, 7, 23, 59, 59, 0, time.UTC), true)...)
	res = append(res, set.Between(
		time.Date(2021, 8, 1, 0, 0, 0, 0, time.UTC),
		time.Date(2021, 8, 7, 23, 59, 59, 0, time.UTC), true)...)

	return res
}

// func main() {
// 	fmt.Println(len(ProjectDailies()))
// 	fmt.Println(len(ProjectMultipleRulesDaily()))
// 	fmt.Println(len(ProjectMultipleRulesDailyOld()))
// 	fmt.Println(len(ProjectWeekly()))
// 	fmt.Println(len(ProjectMultipleRulesWeekly()))
// 	fmt.Println(len(ProjectMultipleRulesWeeklyOld()))
// }

func generateWeek(w http.ResponseWriter, r *http.Request) {
	dates := GenerateEvents()
	// fmt.Printf("%v", dates)
	json.NewEncoder(w).Encode(dates)
}

func handleRequests() {
	http.HandleFunc("/", generateWeek)
	log.Fatal(http.ListenAndServe(":8000", nil))
}

func main() {
	handleRequests()
}

File: node/rrule-gen.js
import smu from "rrule";
const { RRuleSet, RRule } = smu;

export function generateEvents() {
  // 3 days a week
  const rule1 = new RRule(
    {
      freq: RRule.WEEKLY,
      dtstart: new Date(Date.UTC(2021, 1, 1, 10, 30)),
      interval: 1,
      byweekday: [RRule.MO, RRule.WE, RRule.FR],
    },
    true
  );
  // Daily
  const rule2 = new RRule(
    {
      freq: RRule.DAILY,
      dtstart: new Date(Date.UTC(2021, 1, 1, 10, 30)),
    },
    true
  );

  // With exclusions and byweekday
  const rule3 = new RRuleSet(true);
  rule3.rrule(
    new RRule({
      freq: RRule.WEEKLY,
      dtstart: new Date(Date.UTC(2021, 1, 1, 17, 30)),
      interval: 1,
    })
  );

  rule3.exdate(new Date(Date.UTC(2021, 7, 16, 17, 30)));
  rule3.exdate(new Date(Date.UTC(2021, 8, 27, 17, 30)));
  rule3.rdate(new Date(Date.UTC(2021, 7, 5, 10, 30)));
  rule3.rdate(new Date(Date.UTC(2021, 7, 16, 10, 30)));
  rule3.rdate(new Date(Date.UTC(2021, 8, 27, 10, 30)));

  return [
    ...rule1.between(
      new Date(Date.UTC(2021, 7, 1)),
      new Date(Date.UTC(2021, 7, 7, 23, 59, 59)),
      true
    ),
    ...rule2.between(
      new Date(Date.UTC(2021, 7, 1)),
      new Date(Date.UTC(2021, 7, 7, 23, 59, 59)),
      true
    ),
    ...rule3.between(
      new Date(Date.UTC(2021, 7, 1)),
      new Date(Date.UTC(2021, 7, 7, 23, 59, 59)),
      true
    ),
  ];
}

// function generateRandomDate(interval) {
//   const startDate = new Date(2002, 1, 1).valueOf();
//   const timestamp = Math.round(Math.random() * startDate);
//   return [new Date(timestamp), new Date(timestamp + interval)];
// }

// export function projectMaxExpectedEvents() {
//   const iterations = 1000000;
//   let res = [];

//   for (let i = 0; i < iterations; i++) {
//     const [min, max] = generateRandomDate(60 * 60);
//     res = [...res, rule1.between(min, max)];
//   }

//   return res;
// }

const numberOfEvents = 10000;

export function projectDailies() {
  const dtstart = new Date(Date.UTC(2021, 1, 1, 10, 30));
  const rule = new RRule(
    {
      freq: RRule.DAILY,
      dtstart: dtstart,
    },
    true
  );

  const dtend = new Date(dtstart);
  dtend.setDate(dtend.getDate() + numberOfEvents);

  return rule.between(dtstart, dtend, true);
}

export function projectMultipleRulesDaily() {
  const dtstart = new Date(Date.UTC(2021, 1, 1, 10, 30));
  const rule = new RRule(
    {
      freq: RRule.DAILY,
      dtstart: dtstart,
    },
    true
  );
  const iterations = numberOfEvents;
  let res = [];

  for (let i = 0; i < iterations; i++) {
    res = [...res, rule.between(dtstart, dtstart, true)];
  }
  return res;
}

export function projectMultipleRulesDailyOld() {
  const dtstart = new Date(Date.UTC(2021, 1, 1, 10, 30));
  const rule = new RRule(
    {
      freq: RRule.DAILY,
      dtstart: new Date(Date.UTC(2019, 1, 1, 10, 30)),
    },
    true
  );

  const iterations = numberOfEvents;
  let res = [];

  for (let i = 0; i < iterations; i++) {
    res = [...res, rule.between(dtstart, dtstart, true)];
  }
  return res;
}

export function projectWeekly() {
  const dtstart = new Date(Date.UTC(2021, 1, 1, 10, 30));
  const rule = new RRule(
    {
      freq: RRule.WEEKLY,
      dtstart: dtstart,
      interval: 1,
      byweekday: [RRule.MO, RRule.WE, RRule.FR],
    },
    true
  );

  const dtend = new Date(dtstart);
  dtend.setDate(dtend.getDate() + (numberOfEvents / 3) * 7);

  return rule.between(dtstart, dtend, true);
}

export function projectMultipleRulesWeekly() {
  const dtstart = new Date(Date.UTC(2021, 1, 1, 10, 30));
  const rule = new RRule(
    {
      freq: RRule.WEEKLY,
      dtstart: dtstart,
      interval: 1,
      byweekday: [RRule.MO, RRule.WE, RRule.FR],
    },
    true
  );
  const iterations = numberOfEvents / 3;

  const dtend = new Date(dtstart);
  dtend.setDate(dtend.getDate() + 6);

  let res = [];

  for (let i = 0; i < iterations; i++) {
    res = [...res, ...rule.between(dtstart, dtend, true)];
  }
  return res;
}

export function projectMultipleRulesWeeklyAsync() {
  console.log("A");
  return new Promise((resolve) => {
    const dtstart = new Date(Date.UTC(2021, 1, 1, 10, 30));
    const rule = new RRule(
      {
        freq: RRule.WEEKLY,
        dtstart: dtstart,
        interval: 1,
        byweekday: [RRule.MO, RRule.WE, RRule.FR],
      },
      true
    );
    const iterations = numberOfEvents / 3;

    const dtend = new Date(dtstart);
    dtend.setDate(dtend.getDate() + 6);

    let res = [];

    for (let i = 0; i < iterations; i++) {
      res = [...res, ...rule.between(dtstart, dtend, true)];
    }
    console.log("B");
    return resolve(res);
  });
}

export function projectMultipleRulesWeeklyOld() {
  const dtstart = new Date(Date.UTC(2021, 1, 1, 10, 30));
  const rule = new RRule(
    {
      freq: RRule.WEEKLY,
      dtstart: new Date(Date.UTC(2019, 1, 1, 10, 30)),
      interval: 1,
      byweekday: [RRule.MO, RRule.WE, RRule.FR],
    },
    true
  );
  const iterations = numberOfEvents / 3;

  const dtend = new Date(dtstart);
  dtend.setDate(dtend.getDate() + 6);

  let res = [];

  for (let i = 0; i < iterations; i++) {
    res = [...res, ...rule.between(dtstart, dtend, true)];
  }
  return res;
}

// console.log("projectDailies: ", projectDailies().length);
// console.log("projectMultipleRulesDaily: ", projectMultipleRulesDaily().length);
// console.log(
//   "projectMultipleRulesDailyOld: ",
//   projectMultipleRulesDailyOld().length
// );
// console.log("projectWeekly: ", projectWeekly().length);
// console.log(
//   "projectMultipleRulesWeekly: ",
//   projectMultipleRulesWeekly().length
// );
// console.log(
//   "projectMultipleRulesWeeklyOld: ",
//   projectMultipleRulesWeeklyOld().length
// );
